### PR TITLE
Bug 1961852: policy: reduce lock contention on policy object

### PR DESCRIPTION
**Pull request by squeed**
When a policy selects many pods, there can be lock contention. This doesn't need to be the case - all the pod creation wants to do is update a hash.

So, instead, take a read lock and use a sync-safe map instead.

Signed-off-by: Casey Callendrello <cdc@redhat.com>

This is a cherry-pick of upstream PR https://github.com/ovn-org/ovn-kubernetes/pull/2228

### CODE REVIEW

**Comment by squeed** [comment]
test failures seem like total flakes...
/retest

File: go-controller/pkg/ovn/policy.go [hunk 1]
```diff
@@ -679,17 +687,17 @@ func (oc *Controller) handleLocalPodSelectorDelFunc(
...
-	if np.deleted {
+	// If we never saw this pod, short-circuit
+	if _, ok := np.localPods.LoadAndDelete(logicalPort); !ok {
 		return
 	}
 
-	if _, ok := np.localPods[logicalPort]; !ok {
+	np.RLock()
+	defer np.RUnlock()
+	if np.deleted {
```
Comment (msherif1234) [inline go-controller/pkg/ovn/policy.go:697]: what is the use now of the above deleted flag after the switch to sync map ?

Comment (astoycos) [reply go-controller/pkg/ovn/policy.go:697]: I had the same thought... but it accounts for the case where a namespace deletion event is received before the policy deletion event. In that case we have to clean everything up from the namespace code before the policy code executes. see https://github.com/ovn-org/ovn-kubernetes/pull/1665

Comment (astoycos) [reply go-controller/pkg/ovn/policy.go:697]: An unfortunate fact of our controller design..

Comment (squeed) [reply go-controller/pkg/ovn/policy.go:697]: we hold the read lock so we're sure the NP can't be deleted out from under us. The sync map is just for the list of pods.

**Comment by astoycos** [comment]
/lgtm

Fair warning CI has been rough, `ci/prow/e2e-metal-ipi-ovn-dualstack` is permafailing

**Pull request merged**